File: py/gombot_encrypt.py
# -*- encoding: utf-8 -*-

import os, hashlib, hmac, binascii, struct
from pbkdf2 import PBKDF2
from Crypto.Cipher import AES

from gombot_kdf import do as do_kdf

tohex = binascii.hexlify

# AES-CBC, PKCS#5. SJCL.js can do that.

assert AES.block_size == 16

def pkcs5_padding(datalen):
    needed = 16-(datalen%16)
    return chr(needed)*needed

def encrypt(email, password, data):
    authKey, cryptKey = do_kdf(email, password)
    aesKey = PBKDF2(cryptKey, "identity.mozilla.com/gombot/v1/data/AES", 1, 32)
    hmacKey = PBKDF2(cryptKey, "identity.mozilla.com/gombot/v1/data/HMAC", 1, 32)
    print "aesKey", tohex(aesKey)
    print "hmacKey", tohex(hmacKey)
    IV = os.urandom(16)
    print "IV", tohex(IV)
    c = AES.new(aesKey, mode=AES.MODE_CBC, IV=IV)
    padded_data = data + pkcs5_padding(len(data))
    assert len(padded_data)%AES.block_size == 0, (len(padded_data), AES.block_size)
    ct = c.encrypt(padded_data)
    msg = IV+ct
    # [IV, enc(PADDEDDATA), mac(IV+enc(PADDEDDATA))]
    mac = hmac.new(hmacKey, msg, hashlib.sha256).digest()
    print "enc", [tohex(IV), tohex(ct), tohex(mac)]
    msgmac = msg+mac
    return msgmac

def decrypt(email, password, msgmac):
    authKey, cryptKey = do_kdf(email, password)
    aesKey = PBKDF2(cryptKey, "identity.mozilla.com/gombot/v1/data/AES", 1, 32)
    print "aesKey", tohex(aesKey)
    hmacKey = PBKDF2(cryptKey, "identity.mozilla.com/gombot/v1/data/HMAC", 1, 32)
    msg,mac = msgmac[:-32], msgmac[-32:]
    # mac covers everything else: (IV+enc(data+padding))
    if mac != hmac.new(hmacKey, msg, hashlib.sha256).digest():
        raise ValueError("Corrupt encrypted data")
    IV,ct = msg[:16], msg[16:]
    print "dec", [tohex(IV), tohex(ct), tohex(mac)]
    c = AES.new(aesKey, mode=AES.MODE_CBC, IV=IV)
    padded_data = c.decrypt(ct)
    print "padded data", tohex(padded_data)
    pad = ord(padded_data[-1])
    data = padded_data[:-pad]
    return data

if 0:
    msgmac = encrypt(u"foo@example.org", u"password", "data")
    print "msgmac is", tohex(msgmac)
    print decrypt(u"foo@example.org", u"password", msgmac)

if 0:
    print decrypt(u"foo@example.org", u"password",
                  binascii.unhexlify("342d06fafcc71db7a06a123ec0791ff1a7fc62640da43dd5c33eaee83f8f38575fd713528f231dc4ce3e1033e7a2c274ce349a4871eb59d7f3d4feb6ac6fb85d"))

if 0:
    print "IN"
    #print tohex(encrypt(u"foo@example.org", u"password", "data"))
    print "OUT"

    x="6a329494341af362581f7bf4c0577e4e73165ae273aafb2033ec7286727f24694b57122dc84d7832089887f6c9a9f64eca516a7324976b58a93738b956f81c20"
    y="266cb972c2b403031d960074978f1c59c3700021b18f772883dd9abee880ecb27b8545d608257b5e44466abcc0a499de551b60a61e9ff25abe03108f08e2a1fb"

    print decrypt(u"foo@example.org", u"password",
                  binascii.unhexlify(x))

if 1:
    email = u"andré@example.org"
    password = u"pässwörd"
    #authKey, cryptKey = do_kdf(email, password)
    #print "2 authKey :", tohex(authKey)
    #print "2 cryptKey:", tohex(cryptKey)
    data = '{"kéy": "valuë2"}'
    print "data", len(data), tohex(data)
    msg = encrypt(email, password, data)
    print "msg", tohex(msg)

File: js-node/gombot-encrypt.js

var crypto = require("crypto");
var kdf = require("./gombot-kdf.js").gombot_kdf;
var makeSalt = require("./gombot-kdf.js").makeSalt;
var pbkdf2_sha256_sync = require("./gombot-kdf.js").pbkdf2_sha256_sync;

function encrypt(email, password, data) {
    return kdf(email, password)
        .spread(function(authKey, cryptKey) {return cryptKey;})
        .then(function(cryptKey) {
            var aesKey = pbkdf2_sha256_sync(cryptKey, makeSalt("data/AES"), 1, 32);
            //console.log("aesKey", aesKey.toString("hex"));
            var hmacKey = pbkdf2_sha256_sync(cryptKey, makeSalt("data/HMAC"), 1, 32);
            var IV = Buffer(crypto.randomBytes(16), "binary");
            var c = crypto.createCipheriv("aes256", aesKey.toString("binary"), IV);
            var ct = Buffer.concat([IV, 
                                    Buffer(c.update(data), "binary"), 
                                    Buffer(c.final(), "binary")]);
            var h = crypto.createHmac("sha256", hmacKey.toString("binary"));
            h.update(ct.toString("binary"));
            var mac = Buffer(h.digest(), "binary");
            //console.log([IV.toString("hex"), ct.toString("hex"), mac.toString("hex")]);
            return Buffer.concat([ct, mac]);
        })
    ;
}
exports.encrypt = encrypt;

function compare(a, b) { // vaguely constant-time
    if (a.length != b.length)
        return false;
    var reduction = 0;
    for (var i=0; i < a.length; i++)
        reduction |= (a[i] ^ b[i]);
    return reduction == 0;
}
exports.compare = compare;

function decrypt(email, password, msgmac) {
    return kdf(email, password)
        .spread(function(authKey, cryptKey) {return cryptKey;})
        .then(function(cryptKey) {
            var aesKey = pbkdf2_sha256_sync(cryptKey, makeSalt("data/AES"), 1, 32);
            //console.log("aesKey", aesKey.toString("hex"));
            var hmacKey = pbkdf2_sha256_sync(cryptKey, makeSalt("data/HMAC"), 1, 32);
            var h = crypto.createHmac("sha256", hmacKey.toString("binary"));
            h.update(msgmac.slice(0, msgmac.length-32).toString("binary"));
            var expectedHmac = Buffer(h.digest(), "binary");
            var gotHmac = msgmac.slice(msgmac.length-32, msgmac.length);
            if (!compare(expectedHmac, gotHmac))
                throw Error("Corrupt encrypted data");
            var IV = msgmac.slice(0, 16);
            var msg = msgmac.slice(16, msgmac.length-32);
            var c = crypto.createDecipheriv("aes256", aesKey.toString("binary"), IV);
            var data = Buffer.concat([Buffer(c.update(msg.toString("binary")), "binary"), 
                                      Buffer(c.final(), "binary")]);
            return data;
        })
    ;
}
exports.decrypt = decrypt;

function test_one() {
    encrypt("foo@example.org", "password", "data")
        .then(function(msgmac) {console.log(msgmac.toString("hex"));
                                return decrypt("foo@example.org", "password", 
                                               msgmac);})
        .then(function(data) {console.log("data:", data.toString("hex"));
                              if (data.toString("ascii") == "data")
                                  console.log("ok");
                              else
                                  console.log("NOT OK");
                              })
        .then(null, console.log)
    ;
}

function test_one_a() {
    decrypt("foo@example.org", "password",
            Buffer("266cb972c2b403031d960074978f1c59c3700021b18f772883dd9abee880ecb27b8545d608257b5e44466abcc0a499de551b60a61e9ff25abe03108f08e2a1fb", "hex"))
        .then(function(data) {console.log("data:", data.toString("hex"));
                              if (data.toString("ascii") == "data")
                                  console.log("ok");
                              else
                                  console.log("NOT OK");
                              })
        .then(null, console.log)
    ;
}

test_one_a();
